src/typing/nullSafety.ml

```diff
@@ -0,0 +1,1095 @@
+			Make sure nobody tries to access a field on a nullable value
+		*)
+		method private check_field target access p =
+			if self#is_nullable_expr target then
+				self#error ("Cannot access \"" ^ accessed_field_name access ^ "\" of a nullable value.") p;
+			self#check_expr target
+		(**
+			Check constructor invocation: dont' pass nulable values to not-nullable arguments
+		*)
+		method private check_new cls params args p =
+			let ctor =
+				try
+					Some (get_constructor (fun ctor -> apply_params cls.cl_params params ctor.cf_type) cls)
+				with
+					| Not_found -> None
+			in
+			match ctor with
+				| None ->
+					List.iter self#check_expr args
+				| Some (ctor_type, _) ->
+					let rec traverse t =
+						match follow t with
+							| TFun (types, _) -> self#check_args args types
+							| _ -> fail ~msg:"Unexpected constructor type." p __POS__
+					in
+					traverse ctor_type
+		(**
+			Check calls: don't call a nullable value, dont' pass nulable values to not-nullable arguments
+		*)
+		method private check_call callee args =
+			if self#is_nullable_expr callee then
+				self#error "Cannot call a nullable value." callee.epos;
+			self#check_expr callee;
+			match callee.eexpr with
+				(* Handle other calls *)
+				| _ ->
+					match follow callee.etype with
+						| TFun (types, _) ->
+							let fn_name = match (reveal_expr callee).eexpr with
+								| TField (_, access) -> field_name access
+								| TIdent fn_name -> fn_name
+								| TLocal { v_name = fn_name } -> fn_name
+								| _ -> ""
+							in
+							self#check_args ~callee:fn_name args types
+						| _ ->
+							List.iter self#check_expr args
+		(**
+			Check if specified expressions can be passed to a call which expects `types`.
+		*)
+		method private check_args ?(callee="") args types =
+			match (args, types) with
+				| (a :: args, (arg_name, optional, t) :: types) ->
+					if not optional && not (self#can_pass_expr a t a.epos) then begin
+						let fn_str = if callee = "" then "" else " of function \"" ^ callee ^ "\""
+						and arg_str = if arg_name = "" then "" else " \"" ^ arg_name ^ "\"" in
+						self#error ("Cannot pass nullable value to not-nullable argument" ^ arg_str ^ fn_str ^ ".") a.epos
+					end;
+					self#check_expr a;
+					self#check_args ~callee:callee args types;
+				| _ -> ()
+	end
+
+class class_checker cls report is_safe_file  =
+	object (self)
+			val checker = new expr_checker report is_safe_file
+			val mutable initialized_in_constructor = Hashtbl.create 1
+			val mutable constructor_checked = false
+		(**
+			Entry point for checking a class
+		*)
+		method check =
+			if (not cls.cl_extern) && (not cls.cl_interface) then
+				self#check_var_fields;
+			let check_field f =
+				if not (contains_unsafe_meta f.cf_meta) then
+					Option.may checker#check_root_expr f.cf_expr
+			in
+			Option.may checker#check_root_expr cls.cl_init;
+			Option.may (fun field -> Option.may checker#check_root_expr field.cf_expr) cls.cl_constructor;
```

Comment:
Why does this not go through `check_field`?

Reply:
I don't remember )
Fixed in 8485c8c9d3a603293e7377e66f8711dc1b9f8f55